sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -17,15 +17,56 @@
 
 package org.apache.spark.sql.execution.datasources.v2
 
-import org.apache.spark.sql.Strategy
+import org.apache.spark.sql.{execution, Strategy}
+import org.apache.spark.sql.catalyst.expressions.{And, AttributeReference, AttributeSet}
+import org.apache.spark.sql.catalyst.planning.PhysicalOperation
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 import org.apache.spark.sql.execution.SparkPlan
 import org.apache.spark.sql.execution.streaming.continuous.{WriteToContinuousDataSource, WriteToContinuousDataSourceExec}
 
 object DataSourceV2Strategy extends Strategy {
   override def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
-    case r: DataSourceV2Relation =>
-      DataSourceV2ScanExec(r.output, r.source, r.options, r.pushedFilters, r.reader) :: Nil
+    case PhysicalOperation(project, filters, relation: DataSourceV2Relation) =>
+      val projectSet = AttributeSet(project.flatMap(_.references))
+      val filterSet = AttributeSet(filters.flatMap(_.references))
+
+      val projection = if (filterSet.subsetOf(projectSet) &&
+          AttributeSet(relation.output) == projectSet) {
+        // When the required projection contains all of the filter columns and column pruning alone
+        // can produce the required projection, push the required projection.
+        // A final projection may still be needed if the data source produces a different column
+        // order or if it cannot prune all of the nested columns.
+        relation.output
+      } else {
+        // When there are filter columns not already in the required projection or when the required
+        // projection is more complicated than column pruning, base column pruning on the set of
+        // all columns needed by both.
+        (projectSet ++ filterSet).toSeq
+      }
+
+      val reader = relation.newReader
```

Comment:
to confirm, do we have to do operator pushdown twice now? One in the plan visitor to calculate statistics, one here to build the physical plan, right?

Reply:
It will configure two readers. One for the pushdown when converting to a physical plan and one for stats. The stats one should be temporary, though, since we want to address the problem. Configuring two readers instead of one allows us to decouple the problems so we can move forward with pushdown that works like the other data sources.

Reply:
it's nice to decouple the problem and do pushdown during planning, but I feel the cost is too high in this approach. For file-based data sources, we need to query hive metastore to apply partitioning pruning during filter pushdown, and this can be very expensive. Doing it twice looks scaring to me.

cc @gatorsmile @dongjoon-hyun @mallman , please correct me if I have a wrong understanding.

also cc @wzhfy do you have an estimation about how long it takes to move statistics to physical plan?

Reply:
@cloud-fan, there's nothing forcing other data sources to implement the new trait. Other sources can continue to report stats for the entire table and not account for filters (the code assumes that row counts don't change). This just opens the option of reporting stats that are more accurate using the filters and projection that will be pushed.

Ideally, I think that stats-based decisions would happen after pushdown so we get data that is as accurate as possible. But for now, this fixes the regression for v2 sources that happens because we move pushdown to a later step (conversion to physical plan like the other sources).

Reply:
> there's nothing forcing other data sources to implement the new trait ...

hmmm, I'm a little confused here. All v2 data sources would have to apply pushdown twice right? Or are you suggesting we should not migrate file-based data source to data source v2?

Reply:
I didn't realize you were talking about other v2 sources. Yes, two readers would be configured for v2. If you wanted to avoid it, you could cache when pushdown is expensive in the implementation or we could add something else that prevents that case.

We need to do *something* to fix the current behavior of doing pushdown in the optimizer. I'm perfectly happy with less accurate stats for v2 until stats use the physical plan, or a solution like this where pushdown happens twice. I just don't think it is a good idea to continue with the design where the logical plan needs to use the v2 reader APIs. I think we agree that that should happen once, and conversion to physical plan is where it makes the most sense.

Reply:
OK we need to make a decision here:
1. Apply operator pushdown twice(proposed by tihs PR). This moves the pushdown logic to planner which is more ideal and cleaner. The drawback is, before moving statistics to physical plan, we have some duplicated pushdown code in `DataSourceV2Relation` and applying pushdown twice has performance penalty.
2. Apply operator pushdown only once in the planner. Same as 1, it's cleaner. The drawback is, before moving statistics to physical plan, data source v2 can't report statistics after filter.
3. Apply operator pushdown only once in the optimizer(proposed by https://github.com/apache/spark/pull/21319). It has no performance penalty and we can report statistics after filter. The drawback is, before moving statistics to physical plan, we have a temporary `DataSourceReader` in `DataSourceV2Relation`, which is hacky.

The tradeoff is: shall we bear with hacky code and move forward with the data source v2 operator pushdown support? or shall we make the code cleaner and bear with some performance pemalty(apply pushdown twice or not report stats after filter)? or shall we just hold back and think about how to move stats to physical plan?

cc @marmbrus @jose-torres

Reply:
I don't mind either option #1 or #2. #2 is basically what happens for non-v2 data sources right now. Plus, both should be temporary.

I think it is a bad idea to continue with hacky code that uses the reader in the logical plan. It is much cleaner otherwise and we've spend too much time making sure that everything still works. The main example that comes to mind is setting the requested projection and finding out what output is using pushdown. I think hacks are slowing progress on the v2 sources.

Reply:
Yea the second proposal is what happens for the v1 data sources. For file-based data source we kind of pick the third proposal and add an optimizer rule `PruneFileSourcePartitions` to push down some of the filters to data source at the logical phase, to get precise stats.

I'd like to pick from the 2nd and 3rd proposals(the 3rd proposal is also temporary, before we move stats to physical plan). Applying pushdown twice is hard to workaround(need to cache), while we can keep the `PruneFileSourcePartitions` rule to work around the issue in 2nd proposal for file-based data sources.

Let's also get more inputs from other people.

Reply:
I'm not strongly opposed to any of the options, but 2 would be my choice if I had to pick one. A temporary state where functionality is missing is easier to reason about than temporary states where we deliberately impose a fuzzy lifecycle.

Reply:
@rdblue do you have time to prepare a PR for the 2rd proposal? I can do that too if you are busy with other stuff.

Reply:
Yeah, no problem. I can just remove the stats commit from this one.